reflective access to default methods is broken on jdk9

When running spectator on jdk9 we get errors like:

```
Caused by: java.lang.IllegalAccessException: access to public member failed: com.netflix.spectator.api.RegistryConfig.gaugePollingFrequency()Duration/invokeSpecial, from com.netflix.spectator.api.RegistryConfig/2 (unnamed module @36d4b5c)
	at java.base/java.lang.invoke.MemberName.makeAccessException(MemberName.java:914)
	at java.base/java.lang.invoke.MethodHandles$Lookup.checkAccess(MethodHandles.java:2193)
	at java.base/java.lang.invoke.MethodHandles$Lookup.checkMethod(MethodHandles.java:2133)
	at java.base/java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:2282)
	at java.base/java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(MethodHandles.java:2276)
	at java.base/java.lang.invoke.MethodHandles$Lookup.unreflectSpecial(MethodHandles.java:1800)
	at com.netflix.spectator.impl.Config.lambda$createProxy$3(Config.java:131)
	... 50 more
```

This is coming from the reflection code used to try and call default methods on the config interfaces from proxy objects. See also discussion on jdk mailing list:

http://mail.openjdk.java.net/pipermail/mlvm-dev/2016-June/006646.html

--- a/spectator-api/src/main/java/com/netflix/spectator/impl/Config.java
+++ b/spectator-api/src/main/java/com/netflix/spectator/impl/Config.java
@@ -20,7 +20,10 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.management.ManagementFactory;
 import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
 import java.lang.reflect.Proxy;
 import java.time.Duration;
 import java.time.Instant;
@@ -106,6 +109,29 @@ public final class Config {
     }
   }
 
+  private static boolean isJava8() {
+    String version = ManagementFactory.getRuntimeMXBean().getSpecVersion();
+    return version.startsWith("1.8");
+  }
+
+  private static Object invokeDefault(Object proxy, Method method, Object[] args) throws Throwable {
+    // See: https://github.com/Netflix/spectator/issues/425
+    final Class<?> declaringClass = method.getDeclaringClass();
+    if (isJava8()) {
+      final Constructor<MethodHandles.Lookup> constructor = getConstructor();
+      return constructor.newInstance(declaringClass, MethodHandles.Lookup.PRIVATE)
+          .unreflectSpecial(method, declaringClass)
+          .bindTo(proxy)
+          .invokeWithArguments(args);
+    } else {
+      MethodType rt = MethodType.methodType(method.getReturnType());
+      return MethodHandles.lookup()
+          .findSpecial(declaringClass, method.getName(), rt, declaringClass)
+          .bindTo(proxy)
+          .invokeWithArguments(args);
+    }
+  }
+
   /**
    * Create a proxy class that implements the inferface specified.
    *
@@ -121,16 +147,11 @@ public final class Config {
    */
   @SuppressWarnings("unchecked")
   public static <T> T createProxy(Class<T> cls, Function<String, String> props) {
-    final Constructor<MethodHandles.Lookup> constructor = getConstructor();
     final Class<?>[] interfaces = new Class<?>[] {cls};
     return (T) Proxy.newProxyInstance(classLoader(), interfaces, (proxy, method, args) -> {
       final String name = method.getName();
       if (method.isDefault()) {
-        final Class<?> declaringClass = method.getDeclaringClass();
-        return constructor.newInstance(declaringClass, MethodHandles.Lookup.PRIVATE)
-            .unreflectSpecial(method, declaringClass)
-            .bindTo(proxy)
-            .invokeWithArguments(args);
+        return invokeDefault(proxy, method, args);
       } else if ("get".equals(method.getName())) {
         return props.apply((String) args[0]);
       } else {

--- a/spectator-nflx/src/test/java/com/netflix/spectator/nflx/AutoPluginTest.java
+++ b/spectator-nflx/src/test/java/com/netflix/spectator/nflx/AutoPluginTest.java
@@ -1,5 +1,5 @@
-/**
- * Copyright 2015 Netflix, Inc.
+/*
+ * Copyright 2014-2017 Netflix, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -21,10 +21,13 @@ import com.netflix.governator.guice.LifecycleInjector;
 import com.netflix.governator.lifecycle.LifecycleManager;
 import org.apache.commons.configuration.AbstractConfiguration;
 import org.junit.Assert;
+import org.junit.Assume;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
+
+import java.lang.management.ManagementFactory;
 
 @RunWith(JUnit4.class)
 public class AutoPluginTest {
@@ -35,8 +38,19 @@ public class AutoPluginTest {
     cfg.setProperty("spectator.nflx.enabled", "false");
   }
 
+  private static boolean isJava8() {
+    String version = ManagementFactory.getRuntimeMXBean().getSpecVersion();
+    return version.startsWith("1.8");
+  }
+
   @Test
   public void inject() throws Exception {
+    Assume.assumeTrue("requires java 8", isJava8());
+    // On JDK 9:
+    // Caused by: java.lang.ClassNotFoundException: javax.annotation.Resource
+    // at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:582)
+    // at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:185)
+    // at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:496)
     Injector injector = LifecycleInjector.builder()
         .usingBasePackages("com.netflix")
         .build()

